### ISSUE TITLE
[CHAT][EMOTES] A strange issue on ruRu localization.

### ISSUE THREAD

**Issue by markvetwow** [TrinityCore/TrinityCore#19344]
[//]: # (**********************************)
[//]: # (** Fill in the following fields **)
[//]: # (**********************************)

**Description:**

There is a little issue with emotes in chat,so addons like DBM can not trigger them to work properly

**Current behaviour:**

So, current state is a little strange. In chat,there is an excess letter as "s". If you remove a %s in locale_broadcast_text(or even in broadcast_text) and write a name directly, "s" letter don't disappear. :(

**Expected behaviour:**

Emotes must show correctly in **CHAT** window.

**Steps to reproduce the problem:**

1. Launch ruRu client :D. I really don't know.

I will apply screenshot link under this info

**Branch(es):**

3.3.5

**TC rev. hash/commit:** 

TrinityCore rev. 8ca1e2d2d0e5 2017-03-21 01:47:21 -0300 (3.3.5 branch) (Win64, RelWithDebInfo, Static) (worldserver-daemon)

**TDB version:**  335.62

**Operating system:** Win 8.1 x64.

**Screenshot**: [Imgur.com Link](http://i.imgur.com/35Tuisj.jpg)

**Comment by markvetwow** [comment]
guys,i don't know why, but i suppose that happened because in database all NPCs names set in %s value. So,i guess russian localization doesn't support latin symbols(which is logical :D),so,if you remove an "s" letter from database,it'll be working as needed.

[Screenshot on Imgur.com](http://i.imgur.com/3EcNmIz.jpg)

So, i guess all name values in loc8 must setted to % symbol only,without "s" letter.

**Comment by Magnifikator** [comment]
%s means that the core is replacing this by a string (default for c++)
http://www.cplusplus.com/reference/cstdio/printf/

But maybe the replacement and the cyrillic charset doesn't like each other, dont' know.

**Comment by ariel-** [comment]
%s is for printf, yes, however wow texts use $n as placeholder

**Comment by Magnifikator** [comment]
It depends who's using the placeholder. If the core is replacing something it's %s. If the client is replacing than it's $n or similar placeholders.

In the english broadcasttexts there are a lot of %s. So I guess the core is replacing text. 
Unfortunately I can't read russian language, so I didn't understand the problem well.

**Comment by Shauren** [comment]
core never replaces the placeholders (at least it shouldnt, maybe naxx is badly written)

**Comment by Magnifikator** [comment]
SELECT * FROM broadcast_text WHERE MaleText LIKE '%\%s%'
SELECT * FROM trinity_string WHERE content_default LIKE '%\%s%'

**Comment by ariel-** [comment]
Core doesn't replace it, Apparently client replaces %s with source and $n with target (or achievement name); so it would appear the issue is that ruRU client handles replaces differently.

Still, it would be nice to have ruRU sniffs to confirm the %s isn't present in that locale.

**Comment by tkrokli** [comment]
TrinityCore rev. 270a6396783c 2017-04-02 11:14:24 -0300 (3.3.5 branch) (Win64, Release, Static)
TDB 335.62 plus updates up to and including `2017_04_02_01_world.sql`
```sql
SELECT * FROM `broadcast_text` WHERE `MaleText` LIKE '%\\%s%' OR `FemaleText` LIKE '%\\%s%';
```
`Output result: 2276 row(s)`

```sql
SELECT * FROM `trinity_string` WHERE `content_default` LIKE '%\\%s%';
```
`Output result: 354 row(s)`

**Comment by Magnifikator** [comment]
The sad story about broadcast texts and how I see the problem:

Facts:
1. Broadcast texts holds placeholders like %s and $n (name), $c (class) and $g (gender structures).
2. Only the %s placeholders are replaced by the client by the Name of the speaking NPC.

Claim:
All $-placeholders have to parsed and replaced serverside.

Broadcast texts are mostly used by scripts. That's the reason why it's for the client impossible to guess the meaning of $-placeholders.

Script Example:
a) A player opens a door (gameobject).
b) The script let a NPC tell a broadcast text like this:  $n, please welcome in my home.

As you see, it's impossible for the client to know that $n has to be replaced by the name of the player character that was opening the door. That's the nature of serverside scripted context.

Possible solution:
All scripts need to be expanded to save needed values for the placeholders and parse and replace before tell broadcast text.

New Script Example:
a) A player opens a door (gameobject).
b) The script save the name of the player character that activates the door
c) Parse and replace broadcast text
b) The script let a NPC tell the new broadcast text:  Freddy, please welcome in my home.

**Comment by Shauren** [comment]
@Magnifikator you are wrong here, $ placeholders are replaced correctly if you send receiver guid in the packet (when using CreatureAI::Talk you must pass the second arg, confusingly named whisperTarget)

**Comment by Magnifikator** [comment]
That would fit the last theorie is I just haved added in my previous post :-)

**Comment by Magnifikator** [comment]
Ok, I found the problem. It's in smartscripts.

About 50% of the smartscripts that using action_type 1 (Talk) have a target_type 1 (Self). That's obviously wrong if the creature is speaking with a player.

What happens in this case:
1) Talk param 3 (talktarget) is 0: The text appears with placeholder "Welcome $n"
2)  Talk param 3 (talktarget) is 1 (none of the scripts uses this): The text appears with placeholder replacement "Welcome creaturename". This also not want we want.

The only way is to use in the smartsscripts a player as target. But in a script that's may not so easy as it seams.
Following example worked for me: Target type 21 (Closest player in range). 

Tested on NPC 2044 (Forlorn Spirit)

**Comment by tkrokli** [comment]
Looks like you are on to something useful or important there. I can't say for certain.
There is a downside to using Target Type 21 (closest player in range), though:: if you are not playing alone and there is another player standing "inside" the talking NPC, that player could possibly be named in the scripted talk instead of yourself.

**Comment by Magnifikator** [comment]
Yes, type 21 was just a workaround for me to prove that in general it would work fine with broadcast text replacements clientside.

But, as I wrote before it's must be the core that give at the end the correct value by what the placeholders has to be replaced. In this case, the core must give the correct GUID. But, because of the nature of scripted environment it's may be not so easy.

I will show this in an example script:

1. A player open a door and enters a house.
2. script is doing something 
3. script is doing something 
4. a wild battle against monster starts
5. script is doing something 
6. script is doing something 
7. at the end one NPC says "Thank you $n that you come in my house an killed all the monster"

But from where will you know at the end that you need the GUID of player in the first line activating the door ?

I think in some cases you need to save the GUID at a specific point. Maybe I'm thinking also to far away.

**Comment by tkrokli** [comment]
I haven't tested anything related to your findings in your testing, but from previous experience I want to say this:
`smart_scripts.target_type` = `SMART_TARGET_ACTION_INVOKER` (7)
(Unit who caused this Event to occur)
should _in theory_ be the best choice to get the right player's name in a script talk action.
Core scripts should also include targeted `player` as `Talk(player)` target, depending on each script content.

**Comment by Magnifikator** [comment]
I already had tested with 7, was not working. Will write tomorrow a real example.

**Comment by tkrokli** [comment]
I don't know the core coding well enough to point to what goes wrong.
I can only guess something related to target/player selection is broken.

**Comment by Magnifikator** [comment]
Here's a practical example:

Presettings:
- Enable Logger.scripts.ai=2,Console Server in conf file for more details
- Make Alliance character
- .go xyz -9552.34 -1435.39 62.2958 0
- .quest add 75   (prequest, so you can open the chest)
- open the chest (GO 1562, triggers event_script 264 which temporary spawns NPC 2044)
- NPC 2044 (Forlorn Spirit) spawns
- This NPC has smartscripts action_type 1 (Talk) which random takes 3 broadcast texts:
ID 503: Let the legend of Stalvan rest! (no placeholder!)
ID 504: Who is this mere $r that meddles with that which is past?  May the legend of Stalvan die along with you!
ID 505: Toil not in matters of the past, $n!

**Comment by tkrokli** [comment]
Respawning the NPC spawned from a gameobject multiple times is normally not blizzlike (they usually added working fixes to issues like that after any exploits were found on live servers), but TC unfortunately contains many more scripts where you can spawn almost as many spawns as you want (or at least as many as you can trigger in the time it takes to despawn or deactivate the gameobject). The scripts need to be updated to make the objects inactive as long as the spawned NPC is present in the game world, unless this can be done via DB conditions.

**Comment by offl** [comment]
There's an another problem. For Russian and Spanish languages creature/player gender can be important.
For example: On aggro Ethereal Scavenger will say this:
>You're far from home, stranger.

But in Russian or Spanish his quote requires creature/player gender:
>Estás lejos de tu casa, **$g**extranjero:extranjera;.

>Далеко же ты **$g**ушел:ушла; от дома...

So I think **on aggro** creature should always say text to player.

**Comment by Magnifikator** [comment]
The same is in German language. But that's not the problem. For this is the $G (gender) placeholder.

I'm quite sure that, if the GUID is correct, the client will replace $G placeholders correct.
